labs/computing-experiment-subjects/index.md

```diff
@@ -0,0 +1,213 @@
+        WHERE
+            timestamp BETWEEN '{decisions_start}' AND '{decisions_end}'
+""")
+
+spark.sql("SELECT * FROM decisions LIMIT 1")
+```
+
+
+
+
+<table border='1'>
+<tr><th>uuid</th><th>timestamp</th><th>process_timestamp</th><th>visitor_id</th><th>session_id</th><th>account_id</th><th>campaign_id</th><th>experiment_id</th><th>variation_id</th><th>attributes</th><th>user_ip</th><th>user_agent</th><th>referer</th><th>is_holdback</th><th>revision</th><th>client_engine</th><th>client_version</th></tr>
+<tr><td>F4F1EF48-6BC2-4153-A1DA-C29E39B772F9</td><td>2020-05-25 15:27:33.085</td><td>2020-05-25 15:29:21.197</td><td>visitor_1590445653085</td><td>-1235693267</td><td>596780373</td><td>18149940006</td><td>18156943409</td><td>18174970251</td><td>[[100,, browserId, ie], [300,, device, iphone], [600,, source_type, direct], [200,, campaign, frequent visitors], [, ...</td><td>75.111.77.0</td><td>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/53...</td><td>https://app.optimizely.com/</td><td>false</td><td>null</td><td>ricky/fakedata.pwned</td><td>1.0.0</td></tr>
+</table>
+
+
+
+
+## Compute experiment subjects
+
+Next we'll group our decision data by `decision_subject_id` in order to produce a table of _experiment subjects_.
+
+Experimental subjects are the randomization units used in our experiment. In most cases each subject corresponds to a web/app user, but they may also represent teams, user sessions, requests, etc., depending on the design of your experiment. In this example we use visitors (identified by the `visitor_id` field) as our experiment subjects.
+
+This query captures the attribution logic used by [Optimizely Full Stack](https://www.optimizely.com/platform/full-stack/) to isolate the visitors tested in a given experiment:
+- In order to be counted in the analysis, a visitor must have at least one "decision" in the analysis window.
+- The variation and the timestamp assigned to that visitor correspond to the variation in the _first_ decision received from that visitor during the analysis window
+
+**Note:** the visitor counting logic used in [Optimizely Web](https://www.optimizely.com/platform/experimentation/) experiments is slightly different.  See [this article](https://help.optimizely.com/Analyze_Results/How_Optimizely_counts_conversions) for details.
+
+
+```python
+spark.sql(f"""
+    CREATE OR REPLACE TEMPORARY VIEW experiment_subjects AS
+        SELECT
+            {decision_subject_id},
+            experiment_id,
+            variation_id,
+            timestamp
+        FROM (
+            SELECT
+                *,
+                RANK() OVER (PARTITION BY experiment_id, {decision_subject_id} ORDER BY timestamp ASC) AS rnk
+            FROM
+                decisions
+        )
+        WHERE
+            rnk = 1
+        ORDER BY timestamp ASC
+""")
+
+spark.sql("""SELECT * FROM experiment_subjects LIMIT 5""")
+```
+
+
+
+
+<table border='1'>
```

Comment:
As is, this lab seems like it's a "I'll show you this first with an existing data set, so you can do it on your own later", but I wonder if the lab would be more powerful if it were "Follow these steps to compute experiment subjects on your data" (which may have to include instructions on how to get the data to begin with).

Reply:
I included a note above about using this notebook with a custom data directory.  I can add a link to oevents too.  The nice thing is that `oevents` and all of these notebooks will use the `OPTIMIZELY_DATA_DIR` if it is specified, so one you've set that variable you can download data and then analyze it with these notebooks without changing anything.